csrc/exceptions.h

```diff
@@ -0,0 +1,268 @@
+  // Add some new context to the message stack.  The last added context
+  // will be formatted at the end of the context list upon printing.
+  // WARNING: This method is O(n) in the size of the stack, so don't go
+  // wild adding a ridiculous amount of context to error messages.
+  void add_context(std::string msg);
+  const std::string& msg() const {
+    return msg_;
+  }
+  const std::vector<std::string>& context() const {
+    return context_;
+  }
+  const std::string& backtrace() const {
+    return backtrace_;
+  }
+  /// Returns the complete error message, including the source location.
+  /// The returned pointer is invalidated if you call add_context() on
+  /// this object.
+  const char* what() const noexcept override {
+    return what_.c_str();
+  }
+  const void* caller() const noexcept {
+    return caller_;
+  }
+  /// Returns only the error message string, without source location.
+  /// The returned pointer is invalidated if you call add_context() on
+  /// this object.
+  const char* what_without_backtrace() const noexcept {
+    return what_without_backtrace_.c_str();
+  }
+
+ private:
+  void refresh_what();
+  std::string compute_what(bool include_backtrace) const;
+};
+
+[[noreturn]] void nvfCheckFail(
+    const char* func,
+    const char* file,
+    uint32_t line,
+    const std::string& msg);
+
+[[noreturn]] void nvfCheckFail(
+    const char* func,
+    const char* file,
+    uint32_t line,
+    const char* msg);
+
+[[noreturn]] void nvfErrorFail(
+    const char* func,
+    const char* file,
+    uint32_t line,
+    const char* condMsg,
+    const char* userMsg);
+
+[[noreturn]] inline void nvfErrorFail(
+    const char* func,
+    const char* file,
+    uint32_t line,
+    const char* condMsg,
+    CompileTimeEmptyString /*userMsg*/) {
+  nvfCheckFail(func, file, line, condMsg);
+}
+
+[[noreturn]] void nvfErrorFail(
+    const char* func,
+    const char* file,
+    uint32_t line,
+    const char* condMsg,
+    const std::string& userMsg);
+
+template <typename... Args>
+decltype(auto) nvfCheckMsgImpl(const char* /*msg*/, const Args&... args) {
+  return to_str(args...);
+}
+inline const char* nvfCheckMsgImpl(const char* msg) {
+  return msg;
+}
+// If there is just 1 user-provided C-string argument, use it.
+inline const char* nvfCheckMsgImpl(const char* /*msg*/, const char* args) {
+  return args;
+}
+
+} // namespace nvfuser
+
+#define STRINGIZE_IMPL(x) #x
+#define STRINGIZE(x) STRINGIZE_IMPL(x)
+
+#define NVF_ERROR(cond, ...)                                \
+  if ((!(cond))) {                                          \
+    nvfuser::nvfErrorFail(                                  \
+        __FUNCTION__,                                       \
+        __FILE__,                                           \
+        static_cast<uint32_t>(__LINE__),                    \
+        #cond " INTERNAL ASSERT FAILED at " STRINGIZE(__FILE__) ":" STRINGIZE(__LINE__) ", please report a bug to NVFuser. ", \
```

Comment:
Let's be more explicit. Ask to report a bug with repro script on https://github.com/NVIDIA/Fuser/issues